Improve swift style + swiftlint

Bunn opened the pull request:
Task/Issue URL: https://app.asana.com/0/0/1201986385932308/f

**Description**:
* Implement  style guide changes
* Update swift lint
* Fork style from RW repo with some added changes

**Steps to test this PR**

1. Check if the https://github.com/duckduckgo/iOS/blob/develop/styleguide/STYLEGUIDE.md is looking OK
2. This PR should have no changes on the actual product, testing if tests are passing and no issue was introduced should be enough
---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

Bunn committed:
WIP: Swift Style Guide

Bunn committed:
New swift lint rules

Rules that required code changes:
* closure_spacing
* trailing_whitespace
* closure_end_indentation
* overridden_super_call
* fatal_error_message

Bunn committed:
Merge branch 'develop' into feature/bunn/swift-style

# Conflicts:
#	DuckDuckGoTests/AppHTTPSUpgradeStoreTests.swift

Bunn committed:
Fix linter warnings

Bunn committed:
Merge branch 'develop' into feature/bunn/swift-style

Bunn committed:
Merge branch 'develop' into feature/bunn/swift-style

Bunn committed:
Fix linter warning after merging downloads PR

jaceklyp commented on DuckDuckGo/DirectoryMonitor.swift:
```diff
@@ -110,7 +110,7 @@ class DirectoryMonitor {
     
     private func debounceTimerDidFire() {
-        guard case .debounce(let source, let timer) = state else { fatalError() }
+        guard case .debounce(let source, let timer) = state else { fatalError("debounce should have source and timer") }
```
it's super minor, but here we rather expect that in method debounceTimerDidFire the state is actually .debounce rather than source or timer are missing

jaceklyp replied:
it's super minor, but here we rather expect that in method debounceTimerDidFire the state is actually .debounce rather than source or timer are missing

jaceklyp commented on styleguide/STYLEGUIDE.md:
```diff
@@ -1,50 +1,1090 @@
+Remember, structs have [value semantics](https://developer.apple.com/library/mac/documentation/Swift/Conceptual/Swift_Programming_Language/ClassesAndStructures.html#//apple_ref/doc/uid/TP40014097-CH13-XID_144). Use structs for things that do not have an identity. An array that contains [a, b, c] is really the same as another array that contains [a, b, c] and they are completely interchangeable. I ...
+
+Classes have [reference semantics](https://developer.apple.com/library/mac/documentation/Swift/Conceptual/Swift_Programming_Language/ClassesAndStructures.html#//apple_ref/doc/uid/TP40014097-CH13-XID_145). Use classes for things that do have an identity or a specific life cycle. You would model a person as a class because two person objects are two different things. Just because two people have th ...
+
```
is it forked from somewhere? do we need it?

Bunn replied:
@jaceklyp Yep, forked from RW guide: https://github.com/duckduckgo/iOS/blob/02df76396be6400dcfadcb4ba7e390ce17e2cb60/styleguide/STYLEGUIDE.md?plain=1#L1-L2

Are you suggesting to remove the entire Classes and Structures section or just the "Which one to use"?

jaceklyp commented on styleguide/STYLEGUIDE.md:
```diff
@@ -1,50 +1,1090 @@
+Remember, structs have [value semantics](https://developer.apple.com/library/mac/documentation/Swift/Conceptual/Swift_Programming_Language/ClassesAndStructures.html#//apple_ref/doc/uid/TP40014097-CH13-XID_144). Use structs for things that do not have an identity. An array that contains [a, b, c] is really the same as another array that contains [a, b, c] and they are completely interchangeable. I ...
+
+Classes have [reference semantics](https://developer.apple.com/library/mac/documentation/Swift/Conceptual/Swift_Programming_Language/ClassesAndStructures.html#//apple_ref/doc/uid/TP40014097-CH13-XID_145). Use classes for things that do have an identity or a specific life cycle. You would model a person as a class because two person objects are two different things. Just because two people have th ...
+
```
is it forked from somewhere? do we need it?

jaceklyp commented on styleguide/STYLEGUIDE.md:
````diff
@@ -1,50 +1,1090 @@
+}
+```
+
+* Long lines should be wrapped at around 70 characters. A hard limit is intentionally not specified.
````
70 seems very narrow, especially if most of us use big screens

Bunn replied:
@jaceklyp what's your suggestion to replace 70? I don't have strong opinions here.

jaceklyp commented on styleguide/STYLEGUIDE.md:
````diff
@@ -1,50 +1,1090 @@
+}
+```
+
+* Long lines should be wrapped at around 70 characters. A hard limit is intentionally not specified.
````
70 seems very narrow, especially if most of us use big screens

jaceklyp commented on styleguide/STYLEGUIDE.md:
```diff
@@ -1,50 +1,1090 @@
+    self.x = x
+    self.y = y
+    self.radius = radius
+  }
```
I'm not sure about the indentation here

Bunn replied:
@jaceklyp you mean the 2 spaces? 
I'll change to 4.

jaceklyp commented on styleguide/STYLEGUIDE.md:
````diff
@@ -1,50 +1,1090 @@
+Here's an example of a well-styled class definition:
+
+```swift
+class Circle: Shape {
````
final probably, it's a good default

jaceklyp replied:
final probably, it's a good default

jaceklyp commented on styleguide/STYLEGUIDE.md:
````diff
@@ -1,50 +1,1090 @@
+**Preferred**:
+```swift
+var diameter: Double {
+  return radius * 2
````
how about the rule of not using returns for single lines?

jaceklyp replied:
how about the rule of not using returns for single lines?

jaceklyp commented on styleguide/STYLEGUIDE.md:
```diff
@@ -1,50 +1,1090 @@
+  comment: String
+) -> Bool {
+  // reticulate code goes here
+}
```
we're entering "very subjective" ground here, remember having exactly the same discussions in previous company ;) and I believe it's worth discussing within team scope. In such case personally I prefer using:

func reticulateSplines(spline: [Double],
                                      adjustmentFactor: Double ...

jaceklyp commented on styleguide/STYLEGUIDE.md:
```diff
@@ -1,50 +1,1090 @@
+  comment: String
+) -> Bool {
+  // reticulate code goes here
+}
```
we're entering "very subjective" ground here, remember having exactly the same discussions in previous company ;) and I believe it's worth discussing within team scope. In such case personally I prefer using:

```
func reticulateSplines(spline: [Double],
                       adjustmentFactor: Double ...
```

Bunn replied:
Yep, makes sense. For now I changed it to reflect your suggestion, which is also my preferred one.

jaceklyp commented on styleguide/STYLEGUIDE.md:
```diff
@@ -1,50 +1,1090 @@
+## Unit Test Names
 
 * We use the when then convention for test:
 
```
LGTM! Small comment from me: overall thanks for putting it all together, nice work, really! I would like to spend the next quick wins days by introducing SwiftFormat and perhaps somehow merging this list of yours with Airbnb code style guide https://github.com/airbnb/swift so it could benefit in two ways: we would have rules in SwiftFormat defined and mapped to rules in the guide so whenever someone's developing smth, the code could be "fixed" with a single shortcut or pre-commit hook and 2nd during the PRs in comments we could quickly link to certain rules inside the style guide, like this: https://github.com/airbnb/swift#use-anyobject Please let me know what you think ;)

jaceklyp commented on styleguide/STYLEGUIDE.md:
```diff
@@ -1,50 +1,1090 @@
+## Unit Test Names
 
 * We use the when then convention for test:
 
```
LGTM! Small comment from me: overall thanks for putting it all together, nice work, really! I would like to spend the next quick wins days by introducing SwiftFormat and perhaps somehow merging this list of yours with Airbnb code style guide https://github.com/airbnb/swift so it could benefit in two ways: we would have rules in SwiftFormat defined and mapped to rules in the guide so whenever someone's developing smth, the code could be "fixed" with a single shortcut or pre-commit hook and 2nd during the PRs in comments we could quickly link to certain rules inside the style guide, like this: https://github.com/airbnb/swift#use-anyobject Please let me know what you think ;)

Bunn replied:
Replied on MM :)

Bunn committed:
Address PR comments

Bunn committed:
Styleguide changes

Remove Classes and Structures description
Change line wrap to 150 instead of 70

Bunn committed:
Merge branch 'develop' into feature/bunn/swift-style

# Conflicts:
#	DuckDuckGo/TabManager.swift
#	DuckDuckGoTests/AmpMatchingTests.swift
#	DuckDuckGoTests/URLParameterTests.swift

jaceklyp reviewed (approved):
LGTM!
